HDS-710 hero component

NikoHelle opened the pull request:
## Description

New Hero component with multiple variants.

The component type is automatically determined from its children and their order. 

Children can be a
- Hero.Card
- Hero.Image
- Hero.BackgroundImage
- Hero.WideImage

There should be only one `Hero.Card` and one (or zero) type of images. Content can be mixed in any way, but only certain combinations are handled.

If there is a `Hero.Image` before `Hero.Card`, then the image is on the left of course.

If there is a `Hero.BackgroundImage` after a `Hero.Card`, then the version with angled koros with a background image is created.

`Hero.WideImage` should always be after a `Hero.Card`. There is no version where wide image can be on top.

`Hero.Card` can have any kind of children, but usually a heading, paragraphs and buttons are used.

## Related Issue

https://helsinkisolutionoffice.atlassian.net/browse/HDS-710

## Motivation and Context

<!-- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

There are no actual functionalities so only tests are some basic jest tests and visual tests.

harriplappalainen commented on packages/react/src/components/hero/Hero.tsx:
```diff
@@ -0,0 +1,275 @@
+    return imageChildIndex === 0 ? 'imageLeft' : 'imageRight';
+  };
+
+  const type = getHeroType();
```
If `variant` prop is used, I think it would remove the need for figuring out where the elements are, right? This `getHeroType` function is sort of hard to read so it would improve this bit here in my opinion.

NikoHelle replied:
Removed.

harriplappalainen commented on packages/react/src/components/koros/Koros.tsx:
```diff
@@ -36,6 +36,16 @@ export type KorosProps = {
+   * Useful when koros should cover part of the adjacent element
+   * Koros is shifted to the direction of the koros shape
+   */
+  shift?: boolean;
```
I 'm not sure understand what this does 😆 Does this cause the Koros to overlap with an element above/under it? If so, then it needs `flipHorizontal` to be set as well sometimes? And if so, then maybe that could be mentioned.

NikoHelle replied:
Without this the gaps in the Koros shape shows the background color, not the image below/above it. This property pulls (with negative-margin) the adjacent element/image and the element/image is rendered below the Koros.

Could also be done by adding extra elements and setting `position: absolute` to the container of the koros. But anyway the koros height must be calculated and it varies with the shape.

I added the positioning and height adjustment to the Koros component itself, so those are re-usable and not Hero specific.

The new "compact" property cuts the Koros, so it is not  85px high, just exactly the height of the shape/pattern of the Koros.

NikoHelle replied:
Added better descriptions to stories.

minevala replied:
I was also thinking would it make it obsolete if the heights of the koros could be exported somehow 🤔

NikoHelle replied:
Changed

NikoHelle commented on packages/core/src/components/hero/hero.css:
```diff
@@ -0,0 +1,252 @@
+}
+.hds-hero p {
+  font-size: var(--fontsize-body-l);
+}
```
Same reason as above

NikoHelle replied:
Changed

NikoHelle commented on packages/core/src/components/hero/hero.stories.js:
```diff
@@ -0,0 +1,196 @@
+      <div class="hds-hero--with-background__background">
+        ${image}
+      </div>
+   </div>
```
Fixed!

NikoHelle commented on packages/core/src/components/hero/hero.stories.js:
```diff
@@ -0,0 +1,196 @@
+      </div>
+      </div>
+    </div>
+  </div>`;
```
Ran these through a validator and fixed these!

NikoHelle commented on packages/core/src/components/hero/hero.stories.js:
```diff
@@ -0,0 +1,196 @@
+  </style>
+<div class="hds-hero custom-theme hds-hero--angled-koros">
+   <div class="hds-hero--with-background__container">
+   <div class="hds-hero--with-background__contentColumns">
```
That is a modifier specific element, so should be ok.

NikoHelle commented on packages/core/src/components/hero/hero.css:
```diff
@@ -0,0 +1,252 @@
+  .hds-hero__content--two-columns > *:first-child {
+    margin-right: var(--spacing-xl);
+  }
+  .hds-hero__content--two-columns > *:last-child {
```
I don't see the need for those extra classes. Extra clutter, especially when those must also be individually imported to React and applied to child elements.

NikoHelle commented on packages/core/src/components/hero/hero.css:
```diff
@@ -0,0 +1,252 @@
+    width: auto;
+  }
+
+  .hds-hero--with-background__emptyColumn {
```
Ah, yes. Jumping between core and React does this.

NikoHelle commented on packages/core/src/components/hero/hero.css:
```diff
@@ -0,0 +1,252 @@
+  .hds-hero {
+    --padding-horizontal: var(--horizontal-padding-x-large);
+  }
+  .hds-hero h1 {
```
Same reason as above.

NikoHelle commented on packages/core/src/components/hero/hero.css:
```diff
@@ -0,0 +1,252 @@
+    --padding-horizontal: var(--horizontal-padding-large);
+    --padding-vertical: var(--spacing-layout-xl);
+  }
+  .hds-hero p {
```
Same reason as above

NikoHelle replied:
Changed.

NikoHelle commented on packages/core/src/components/hero/hero.css:
```diff
@@ -0,0 +1,252 @@
+    z-index: 1;
+    height: 100%;
+    inset: var(--angled-koros-inset);
+    overflow: visible;
```
I'll wait for the tool to do this. Merging can't be done until css linting is fixed - when the tool is in place.

NikoHelle commented on packages/react/src/components/hero/Hero.module.scss:
```diff
@@ -0,0 +1,113 @@
+@value heroCss: "../../../../core/src/components/hero/hero.css";
+
+.hero {
+  composes: hds-hero from heroCss;
```
Hmmm.... is it safer to repeat hard-to-read strings? If "../../../../" is repeated 23 times and one comma is missing some where, it is harder to detect.

harriplappalainen replied:
I don't understand what you mean but my main concern is that here the value is picked from outside react package. Elsewhere it's picked straight from hds-core dependency when composing core classes. I worry if there are some cases where this won't work.

NikoHelle replied:
Now I see what you see 😄 I'll change the path to the core

NikoHelle commented on packages/react/src/components/hero/Hero.stories.tsx:
```diff
@@ -0,0 +1,502 @@
+  imagePosition: {
+    options: ['right', 'left'],
+    control: { type: 'radio' },
+    defaultValue: 'right',
```
Ah, good idea. Didn't see that anywhere

NikoHelle replied:
Added to multiple places when a non-component prop is used.

NikoHelle commented on packages/react/src/components/hero/Hero.stories.tsx:
```diff
@@ -0,0 +1,502 @@
+  paragraphPadding: {
+    defaultValue: '0%',
+    control: 'text',
+    description: 'Padding of the p element',
```
Done.

NikoHelle commented on packages/react/src/components/hero/Hero.tsx:
```diff
@@ -0,0 +1,275 @@
+    components: [],
+  };
+
+  getChildrenAsArray(children).forEach((child, index) => {
```
Removed.

NikoHelle commented on packages/react/src/components/hero/Hero.tsx:
```diff
@@ -0,0 +1,275 @@
+
+WideImage.componentName = 'WideImage';
+
+const BackgroundImage = (props: ImgElementAttributes) => {
```
Removed.

NikoHelle commented on packages/core/src/components/hero/hero.css:
```diff
@@ -0,0 +1,252 @@
+  font-weight: normal;
+  line-height: 1em;
+  margin-top: 0;
+}
```
Maybe just Hero.Title and Hero.Text. But forcing devs to use those seems useless. What is the gain?

harriplappalainen replied:
Main gain is just small optimisation because of browser selector matching from right to left (first match all h1/p elements, then the parent class).

Here's some other thoughts:
1. Hero.Title and Hero.Text give users clear elements on what to use and what works here
2. Guiding users to use h1 and p tags here sort of gives the impression that you can put any HTML here or use h2 or spans instead. The usable elements could be mentioned in docs but I don't know if that's as powerful as own specific components for them.
3. No other core component uses heading or p tags, maybe because of the browser selector matching thing. Can @minevala confirm or speculate on this? 😄

NikoHelle replied:
`Hero.Title` would also hide the element type and devs should always check docs/source what's the element so the headings are in the correct order.

NikoHelle replied:
Of course `Hero.H1` can be done, but still seems silly.

minevala commented on packages/react/src/components/hero/Hero.stories.tsx:
```diff
@@ -0,0 +1,498 @@
+  },
+};
+
+export const PlaygroundForDiagonalKoros = (args) => (
```
Do we need a loki test for broken hero?
What is the meaning of the broken hero component?

NikoHelle replied:
Koros heights are exported with `getShapeHeight` in Koros.tsx.

I can make this Hero specific feature and and remove it from Koros.

NikoHelle replied:
Removed the intentionally broken version. Removed  changes in Koros component, except getShapeHeight

minevala reviewed (commented):
Nice work, the examples seem to work pretty well 👍 
Some suggestions, questions and comments

laurakarhu commented:
Otherwise looks great 
https://city-of-helsinki.github.io/hds-demo/hero/?path=/story/components-hero--playground-for-koros&args=flipHorizontal:false;dense:true has the ghost line under the koro.

NikoHelle commented:
> Otherwise looks great https://city-of-helsinki.github.io/hds-demo/hero/?path=/story/components-hero--playground-for-koros&args=flipHorizontal:false;dense:true has the ghost line under the koro. At least with dense variation of basic koro. 

I can't see this with Chrome/FF/Safari on Mac. Anyone else seeing it?

harriplappalainen commented:
> > Otherwise looks great https://city-of-helsinki.github.io/hds-demo/hero/?path=/story/components-hero--playground-for-koros&args=flipHorizontal:false;dense:true has the ghost line under the koro. At least with dense variation of basic koro.
> 
> I can't see this with Chrome/FF/Safari on Mac. Anyone else seeing it?

On big screen it's not there but zoom in and I can actually see two ghost lines above and below the Koro.

NikoHelle commented on packages/core/src/components/hero/hero.css:
```diff
@@ -0,0 +1,252 @@
+  }
+
+  .hds-hero--background-image .hds-hero__content {
+    --padding-vertical: var(--spacing-2-xl);
```
Yes, the box in the background image -variant is aligned differently with padding.

> I was just thinking should the overrides happen on the same hds-hero level as is done in media queries

I'm sure what that line means :)

minevala replied:
I meant that should it be enough to use
```css
.hds-hero--background-image  {
```
 😁

NikoHelle replied:
No, it should affect only the `.hds-hero__content`. If it set to `.hds-hero--background-image` it will affect also `.hds-hero--background-image .hds-hero--with-background__container`

NikoHelle commented on packages/core/src/components/hero/hero.css:
```diff
@@ -0,0 +1,252 @@
+  padding: var(--padding-vertical) var(--padding-horizontal);
+}
+.hds-hero--diagonal-koros .hds-hero--diagonal-koros__koros-and-background {
+  transform: scaleY(-1);
```
Yes, but if I use props, the koros sets its own transform-object to style-attribute, which overrides all transformations set with css classes

NikoHelle replied:
> The sub components might be more flexible because then the lib users can add additional styles via className etc

Hmm... Is 

` <Hero.Title className="custom-class" heading-level={2}>Title</Hero.Title>`
 
more flexible than
 
` <h2 className="custom-class">Title</h2>`
 

The consensus is more with adding `Hero.Title` and `Hero.Text` than using plain HTML, so I'll add them. 

There are no layouts for other headings than h1, so adding heading-level={2} just adds plain h2 without styling.

minevala replied:
I remember a case where there was a need to configure the heading-levels, at least for the screen-readers
But that would require that the heading is not an h1, h2, or anything like that 🤔
Maybe I'm overthinking this.

NikoHelle replied:
Done.

NikoHelle commented on packages/core/src/components/hero/hero.css:
```diff
@@ -0,0 +1,252 @@
+  position: relative;
+  z-index: 1;
+  background-color: var(--background-color);
+}
```
Yes, I'll wait for the linter PR to be ready so it can fix this, like with sorting rules alphabetically.

NikoHelle commented:
Some kind of rounding problem with Chrome, vectors and setting element height.

NikoHelle commented:
A bigger update:
- Hero.Title & Hero.Text added
- removed most of the Koros changes. Included same in the Hero itself.
- simplified Koros usage, it is now absolutely positioned. Needed to add container elements, but reduced number of needed Koros elements
- added stories

NikoHelle commented on packages/core/src/components/hero/hero.css:
```diff
@@ -0,0 +1,252 @@
+
+.hds-hero--with-background__contentColumns {
+  display: flex;
+  max-width: 1200px;
```
Done

harriplappalainen reviewed (commented):
Great improvements! This is a more streamlined version from before, I like it. I had a couple of comments, mainly about Storybook experience.

NikoHelle commented on packages/react/src/components/hero/Hero.stories.tsx:
```diff
@@ -0,0 +1,546 @@
+  ...getDisabledControl('imageSrc'),
+};
+
+export const WithoutImage2 = (args) => {
```
I tried to came up with something better, but story names cannot (there is no much space) be very long. Descriptive name would be "No image, bottom koros overlaying space below" 😄  Heroes are pretty much just visual sugar, so the story name does not say much, but devs can see the component and see what the story is about.

The story is a good one, because it shows the settings for that version of that variant.  There are three different kind of non-image heroes created with `{ variant: "noImage" }`

There is a playground for theme at the moment. Is it wrongly built? It is good to have all args in the story props, because koros and theme settings work hand in hand.

NikoHelle commented on packages/react/src/components/hero/Hero.stories.tsx:
```diff
@@ -0,0 +1,546 @@
+    theme: { ...imageLeftOrRightTheme, ...args.theme },
+    imageSrc: args.imageSrc || imageFile,
+    variant: 'imageLeft',
+  };
```
Stories show each variant, so there should be no need to change the variant, when the story is build just for that variant? `ImageLeft` and `ImageRight` used to be in one story, but separate stories give better user experience, because story controls are not very convenient. And then loki images are created for each variant.

All props, except the variant, can be set in the story controls. The story args are passed to the hero component. The props object is created inside the story, but args are set to it. So user can override them.

```
const heroProps: HeroProps = {
 koros: **args.koros**,
centeredContent: **args.centeredContent,**
 theme: { ...imageLeftOrRightTheme, ...**args.theme** },
 imageSrc: **args.imageSrc** || imageFile,
variant: 'imageLeft',
};
```

[Example](https://city-of-helsinki.github.io/hds-demo/hero/?path=/story/components-hero--image-left&args=centeredContent:true;koros.dense:true;theme.--background-color:!hex(ff9)) For some reason all props are not stored to urls, like imageSrc.

NikoHelle replied:
Good point, I'll make `variant` editable, so users can see the value immediately. 

The [Playground for theme](https://city-of-helsinki.github.io/hds-demo/hero/?path=/story/components-hero--playground-for-theme) outputs the applied theme under the component so it is easier to copy-paste.

Could that be useful with all props too? A common component that would dynamically show how to use the component with currently applied props?

harriplappalainen replied:
>Could that be useful with all props too? A common component that would dynamically show how to use the component with currently applied props?

The storybook Controls section does that job nicely by default so I don't think it's needed. It's pretty handy to display the theme object though.

NikoHelle replied:
All properties of all stories are now enabled and there are no props set in the code.

harriplappalainen commented on packages/react/src/components/hero/Hero.stories.tsx:
```diff
@@ -0,0 +1,546 @@
+    control: 'text',
+    description: 'Horizontal padding on large screens >=992px.',
+  },
+  horizontalPaddingXLarge: {
```
Add the notion about these being strictly for Storybook, not actual props.
Reminder: `description: 'Storybook control, not an actual component property.'`

harriplappalainen replied:
Or add something about this being an actual theme variable, like: `description: 'Storybook control for theme variable, not an actual component property.'` ?

NikoHelle replied:
Ok

NikoHelle replied:
New descriptions and summaries have been set.

harriplappalainen commented on packages/react/src/components/hero/Hero.stories.tsx:
```diff
@@ -0,0 +1,546 @@
+    control: 'boolean',
+    description: 'Use dense koros version or not',
+  },
+  diagonalKorosPosition: korosPosition,
```
Hide, dense and flipHorizontal are Koros props but I don't know if that needs to mentioned. Maybe add the notion about this `diagonalKorosPosition` not being an actual component prop?

NikoHelle replied:
I'll mention it is set via theme.

NikoHelle replied:
New descriptions and summaries have been set.

harriplappalainen reviewed (approved):
I'm happy with this 🌞 

EDIT: Oh but the accessibility audit is still needed.

NikoHelle commented:
And style linter fixes.

NikoHelle commented:
- [ ] Accessibility audit
- [ ] Apply style linter checks when linter is ready

laurakarhu reviewed (changes requested):
<img width="734" alt="image" src="https://github.com/City-of-Helsinki/helsinki-design-system/assets/75484068/07c8ac72-aa14-4252-9a99-9ebfb30f5bd4">
<img width="545" alt="image" src="https://github.com/City-of-Helsinki/helsinki-design-system/assets/75484068/52d9a8f8-ffee-479c-8d4d-ceb19af9e462">

Remarks from accessibility audit
- [x] focus on hero buttons looks different on navigating with tab and with mouse
- [ ] buttons get now blue text on focus and especially on black background it is impossible to read, but same blue appears also on other backgrounds when using keyboard, see images above
